File: src/services/AdminService.js
import {
  collection,
  getDocs,
  query,
  where,
  doc,
  updateDoc,
  deleteDoc,
  getDoc,
  serverTimestamp,
  addDoc,
} from 'firebase/firestore';
import { db } from '../config/firebase';

/**
 * AdminService - Handles admin panel operations
 */
export const AdminService = {
  /**
   * Check if current user is admin
   */
  async isAdmin(userId) {
    try {
      const userDoc = await getDoc(doc(db, 'users', userId));
      if (!userDoc.exists()) return false;
      return userDoc.data().isAdmin === true;
    } catch (error) {
      console.error('Error checking admin status:', error);
      return false;
    }
  },

  /**
   * Get all users
   */
  async getAllUsers() {
    try {
      const usersRef = collection(db, 'users');
      const snapshot = await getDocs(usersRef);

      const users = snapshot.docs.map((doc) => ({
        id: doc.id,
        ...doc.data(),
      }));

      // Filter out any deleted users from the view (optional, comment this out if you want to see deleted users)
      // const activeUsers = users.filter(u => u.accountStatus !== 'deleted');

      // Sort by creation date (newest first) with fallback for missing dates
      users.sort((a, b) => {
        const timeA = a.createdAt ? (a.createdAt.toDate?.() || new Date(a.createdAt)).getTime() : 0;
        const timeB = b.createdAt ? (b.createdAt.toDate?.() || new Date(b.createdAt)).getTime() : 0;
        return timeB - timeA;
      });

      console.log(`Admin: Fetched ${users.length} users from Firestore`);
      return { success: true, users };
    } catch (error) {
      console.error('Error getting users:', error);
      return { success: false, error: error.message };
    }
  },

  /**
   * Get all reports
   */
  async getAllReports() {
    try {
      const reportsRef = collection(db, 'reports');
      const snapshot = await getDocs(reportsRef);

      const reports = snapshot.docs.map((doc) => ({
        id: doc.id,
        ...doc.data(),
      }));

      // Sort by creation date (newest first)
      reports.sort((a, b) => {
        const timeA = a.createdAt?.toDate?.().getTime() || new Date(a.createdAt).getTime();
        const timeB = b.createdAt?.toDate?.().getTime() || new Date(b.createdAt).getTime();
        return timeB - timeA;
      });

      return { success: true, reports };
    } catch (error) {
      console.error('Error getting reports:', error);
      return { success: false, error: error.message };
    }
  },

  /**
   * Get all verification requests (pending only by default)
   */
  async getVerificationRequests(includeAll = false) {
    try {
      const verificationsRef = collection(db, 'verifications');
      let q;
      
      if (includeAll) {
        q = query(verificationsRef);
      } else {
        q = query(verificationsRef, where('status', '==', 'pending'));
      }
      
      const snapshot = await getDocs(q);

      const verifications = [];
      
      for (const verDoc of snapshot.docs) {
        const verData = verDoc.data();
        
        // Get user details
        const userDoc = await getDoc(doc(db, 'users', verData.userId));
        const userData = userDoc.exists() ? userDoc.data() : null;
        
        verifications.push({
          id: verDoc.id,
          ...verData,
          userName: userData?.name || 'Unknown',
          userEmail: userData?.email || '',
          userPhotos: userData?.photos || []
        });
      }

      // Sort by creation date (oldest pending first)
      verifications.sort((a, b) => {
        const timeA = a.submittedAt?.toDate?.().getTime() || new Date(a.submittedAt).getTime();
        const timeB = b.submittedAt?.toDate?.().getTime() || new Date(b.submittedAt).getTime();
        return timeA - timeB;
      });

      return { success: true, verifications };
    } catch (error) {
      console.error('Error getting verification requests:', error);
      return { success: false, error: error.message };
    }
  },

  /**
   * Approve user verification (enhanced to work with VerificationService)
   */
  async approveVerification(verificationId, adminId) {
    try {
      const verificationRef = doc(db, 'verifications', verificationId);
      const verificationDoc = await getDoc(verificationRef);
      
      if (!verificationDoc.exists()) {
        return { success: false, error: 'Verification not found' };
      }

      const verData = verificationDoc.data();

      // Update verification status
      await updateDoc(verificationRef, {
        status: 'approved',
        reviewedAt: serverTimestamp(),
        reviewedBy: adminId,
        rejectionReason: null
      });

      // Update user profile based on verification type
      const userRef = doc(db, 'users', verData.userId);
      const updateData = {
        [`verification.${verData.verificationType}`]: 'approved',
        [`verification.${verData.verificationType}VerifiedAt`]: serverTimestamp(),
        updatedAt: serverTimestamp()
      };

      // Set the specific verified flag
      if (verData.verificationType === 'phone') {
        updateData.isPhoneVerified = true;
      } else if (verData.verificationType === 'photo') {
        updateData.isPhotoVerified = true;
      } else if (verData.verificationType === 'id') {
        updateData.isIDVerified = true;
      }

      await updateDoc(userRef, updateData);

      return { success: true };
    } catch (error) {
      console.error('Error approving verification:', error);
      return { success: false, error: error.message };
    }
  },

  /**
   * Reject user verification (enhanced)
   */
  async rejectVerification(verificationId, adminId, reason = 'Does not meet verification requirements') {
    try {
      const verificationRef = doc(db, 'verifications', verificationId);
      const verificationDoc = await getDoc(verificationRef);
      
      if (!verificationDoc.exists()) {
        return { success: false, error: 'Verification not found' };
      }

      const verData = verificationDoc.data();

      // Update verification status
      await updateDoc(verificationRef, {
        status: 'rejected',
        rejectionReason: reason,
        reviewedAt: serverTimestamp(),
        reviewedBy: adminId
      });

      // Update user profile
      const userRef = doc(db, 'users', verData.userId);
      await updateDoc(userRef, {
        [`verification.${verData.verificationType}`]: 'rejected',
        [`verification.${verData.verificationType}RejectionReason`]: reason,
        updatedAt: serverTimestamp()
      });

      return { success: true };
    } catch (error) {
      console.error('Error rejecting verification:', error);
      return { success: false, error: error.message };
    }
  },

  /**
   * Suspend user account
   */
  async suspendUser(userId, reason = '') {
    try {
      const userRef = doc(db, 'users', userId);
      await updateDoc(userRef, {
        accountStatus: 'suspended',
        suspensionReason: reason,
        suspendedAt: serverTimestamp(),
      });

      return { success: true };
    } catch (error) {
      console.error('Error suspending user:', error);
      return { success: false, error: error.message };
    }
  },

  /**
   * Unsuspend user account
   */
  async unsuspendUser(userId) {
    try {
      const userRef = doc(db, 'users', userId);
      await updateDoc(userRef, {
        accountStatus: 'active',
        updatedAt: serverTimestamp(),
      });

      return { success: true };
    } catch (error) {
      console.error('Error unsuspending user:', error);
      return { success: false, error: error.message };
    }
  },

  /**
   * Manually verify a user (admin action)
   */
  async verifyUser(userId) {
    try {
      console.log('=== VERIFY USER START ===');
      console.log('Verifying user:', userId);
      
      // Get user data first
      const userDoc = await getDoc(doc(db, 'users', userId));
      if (!userDoc.exists()) {
        console.error('User not found:', userId);
        return { success: false, error: 'User not found' };
      }
      
      const userData = userDoc.data();
      console.log('User data loaded:', userData);
      
      // Update user profile
      console.log('Updating user profile...');
      const userRef = doc(db, 'users', userId);
      await updateDoc(userRef, {
        isVerified: true,
        'verification.isVerified': true,
        'verification.verifiedAt': serverTimestamp(),
        'verification.verifiedBy': 'admin',
        verificationLevel: 'verified',
        updatedAt: serverTimestamp(),
      });
      console.log('User profile updated successfully');

      // Create a verification record for admin's action
      console.log('Creating verification record...');
      const verificationsRef = collection(db, 'verifications');
      const verificationDoc = await addDoc(verificationsRef, {
        userId: userId,
        verificationType: 'admin_manual',
        status: 'approved',
        submittedAt: serverTimestamp(),
        reviewedAt: serverTimestamp(),
        reviewedBy: 'admin',
        userName: userData.displayName || userData.name || userData.email || 'User',
        userEmail: userData.email || '',
        notes: 'Manually verified by admin'
      });
      console.log('Verification record created with ID:', verificationDoc.id);
      console.log('=== VERIFY USER SUCCESS ===');

      return { success: true };
    } catch (error) {
      console.error('=== VERIFY USER ERROR ===');
      console.error('Error verifying user:', error);
      console.error('Error code:', error.code);
      console.error('Error message:', error.message);
      return { success: false, error: error.message };
    }
  },

  /**
   * Remove verification from a user (admin action)
   */
  async unverifyUser(userId) {
    try {
      // Update user profile
      const userRef = doc(db, 'users', userId);
      await updateDoc(userRef, {
        isVerified: false,
        'verification.isVerified': false,
        verificationLevel: null,
        updatedAt: serverTimestamp(),
      });

      // Optionally, update the verification record to show it was revoked
      const verificationsRef = collection(db, 'verifications');
      const q = query(verificationsRef, where('userId', '==', userId), where('status', '==', 'approved'));
      const snapshot = await getDocs(q);
      
      // Update all approved verifications for this user to 'revoked'
      const updatePromises = snapshot.docs.map(verDoc => 
        updateDoc(doc(db, 'verifications', verDoc.id), {
          status: 'revoked',
          revokedAt: serverTimestamp(),
          revokedBy: 'admin'
        })
      );
      await Promise.all(updatePromises);

      return { success: true };
    } catch (error) {
      console.error('Error removing verification:', error);
      return { success: false, error: error.message };
    }
  },

  /**
   * Delete user account (hard delete)
   */
  async deleteUser(userId) {
    try {
      // Mark as deleted instead of hard delete to maintain data integrity
      const userRef = doc(db, 'users', userId);
      await updateDoc(userRef, {
        accountStatus: 'deleted',
        deletedAt: serverTimestamp(),
      });

      return { success: true };
    } catch (error) {
      console.error('Error deleting user:', error);
      return { success: false, error: error.message };
    }
  },

  /**
   * Update report status
   */
  async updateReportStatus(reportId, status, adminNotes = '') {
    try {
      const reportRef = doc(db, 'reports', reportId);
      await updateDoc(reportRef, {
        status,
        adminNotes,
        resolvedAt: status !== 'pending' ? serverTimestamp() : null,
      });

      return { success: true };
    } catch (error) {
      console.error('Error updating report:', error);
      return { success: false, error: error.message };
    }
  },

  /**
   * Get user statistics
   */
  async getUserStats() {
    try {
      const usersRef = collection(db, 'users');
      const snapshot = await getDocs(usersRef);

      let totalUsers = 0;
      let activeUsers = 0;
      let suspendedUsers = 0;
      let deletedUsers = 0;
      let verifiedUsers = 0;
      let subscribedUsers = 0;

      snapshot.docs.forEach((doc) => {
        const data = doc.data();
        totalUsers++;
        if (data.accountStatus === 'active') activeUsers++;
        if (data.accountStatus === 'suspended') suspendedUsers++;
        if (data.accountStatus === 'deleted') deletedUsers++;
        if (data.isVerified) verifiedUsers++;
        if (data.hasActiveSubscription) subscribedUsers++;
      });

      return {
        success: true,
        stats: {
          totalUsers,
          activeUsers,
          suspendedUsers,
          deletedUsers,
          verifiedUsers,
          subscribedUsers,
        },
      };
    } catch (error) {
      console.error('Error getting user stats:', error);
      return { success: false, error: error.message };
    }
  },
};

export default AdminService;

File: lib/data/services/admin_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import '../models/user_model.dart';

/// AdminService - Handles admin panel operations
/// Ported from web app AdminService.js
class AdminService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  /// Check if current user is admin
  Future<bool> isAdmin(String userId) async {
    try {
      final userDoc = await _firestore.collection('users').doc(userId).get();
      if (!userDoc.exists) return false;
      return userDoc.data()?['isAdmin'] == true;
    } catch (e) {
      return false;
    }
  }

  /// Get all users
  Future<List<UserModel>> getAllUsers() async {
    try {
      final snapshot = await _firestore.collection('users').get();
      
      final users = snapshot.docs
          .map((doc) => UserModel.fromFirestore(doc))
          .toList();

      // Sort by creation date (newest first)
      users.sort((a, b) {
        final timeA = a.createdAt ?? DateTime.now();
        final timeB = b.createdAt ?? DateTime.now();
        return timeB.compareTo(timeA);
      });

      return users;
    } catch (e) {
      return [];
    }
  }

  /// Get all reports
  Future<List<Map<String, dynamic>>> getAllReports() async {
    try {
      final snapshot = await _firestore.collection('reports').get();

      final reports = snapshot.docs.map((doc) {
        final data = doc.data();
        return {
          'id': doc.id,
          ...data,
        };
      }).toList();

      // Sort by creation date (newest first)
      reports.sort((a, b) {
        final timeA = (a['createdAt'] as Timestamp?)?.toDate() ?? DateTime.now();
        final timeB = (b['createdAt'] as Timestamp?)?.toDate() ?? DateTime.now();
        return timeB.compareTo(timeA);
      });

      return reports;
    } catch (e) {
      return [];
    }
  }

  /// Get all verification requests
  Future<List<Map<String, dynamic>>> getVerificationRequests({bool includeAll = false}) async {
    try {
      Query query = _firestore.collection('verifications');
      
      if (!includeAll) {
        query = query.where('status', isEqualTo: 'pending');
      }
      
      final snapshot = await query.get();
      
      final verifications = <Map<String, dynamic>>[];
      
      for (final verDoc in snapshot.docs) {
        final verData = verDoc.data() as Map<String, dynamic>;
        
        // Get user details
        final userId = verData['userId'] as String?;
        Map<String, dynamic>? userData;
        
        if (userId != null) {
          final userDoc = await _firestore.collection('users').doc(userId).get();
          if (userDoc.exists) {
            userData = userDoc.data();
          }
        }
        
        verifications.add({
          'id': verDoc.id,
          ...verData,
          'userName': userData?['displayName'] ?? 'Unknown',
          'userEmail': userData?['email'] ?? '',
          'userPhotos': userData?['photos'] ?? [],
        });
      }

      // Sort by submission date (oldest pending first)
      verifications.sort((a, b) {
        final timeA = (a['submittedAt'] as Timestamp?)?.toDate() ?? DateTime.now();
        final timeB = (b['submittedAt'] as Timestamp?)?.toDate() ?? DateTime.now();
        return timeA.compareTo(timeB);
      });

      return verifications;
    } catch (e) {
      return [];
    }
  }

  /// Approve user verification
  Future<bool> approveVerification(String verificationId, String adminId) async {
    try {
      final verificationRef = _firestore.collection('verifications').doc(verificationId);
      final verificationDoc = await verificationRef.get();
      
      if (!verificationDoc.exists) return false;

      final verData = verificationDoc.data()!;

      // Update verification status
      await verificationRef.update({
        'status': 'approved',
        'reviewedAt': FieldValue.serverTimestamp(),
        'reviewedBy': adminId,
        'rejectionReason': null,
      });

      // Update user profile based on verification type
      final userRef = _firestore.collection('users').doc(verData['userId']);
      final updateData = <String, dynamic>{
        'verification.${verData['verificationType']}': 'approved',
        'verification.${verData['verificationType']}VerifiedAt': FieldValue.serverTimestamp(),
        'updatedAt': FieldValue.serverTimestamp(),
      };

      // Set the specific verified flag
      if (verData['verificationType'] == 'phone') {
        updateData['isPhoneVerified'] = true;
      } else if (verData['verificationType'] == 'photo') {
        updateData['isPhotoVerified'] = true;
      } else if (verData['verificationType'] == 'id') {
        updateData['isIDVerified'] = true;
      }

      await userRef.update(updateData);

      return true;
    } catch (e) {
      return false;
    }
  }

  /// Reject user verification
  Future<bool> rejectVerification(
    String verificationId, 
    String adminId, 
    {String reason = 'Does not meet verification requirements'}
  ) async {
    try {
      final verificationRef = _firestore.collection('verifications').doc(verificationId);
      final verificationDoc = await verificationRef.get();
      
      if (!verificationDoc.exists) return false;

      final verData = verificationDoc.data()!;

      // Update verification status
      await verificationRef.update({
        'status': 'rejected',
        'rejectionReason': reason,
        'reviewedAt': FieldValue.serverTimestamp(),
        'reviewedBy': adminId,
      });

      // Update user profile
      final userRef = _firestore.collection('users').doc(verData['userId']);
      await userRef.update({
        'verification.${verData['verificationType']}': 'rejected',
        'verification.${verData['verificationType']}RejectionReason': reason,
        'updatedAt': FieldValue.serverTimestamp(),
      });

      return true;
    } catch (e) {
      return false;
    }
  }

  /// Suspend user account
  Future<bool> suspendUser(String userId, {String reason = ''}) async {
    try {
      await _firestore.collection('users').doc(userId).update({
        'accountStatus': 'suspended',
        'suspensionReason': reason,
        'suspendedAt': FieldValue.serverTimestamp(),
      });
      return true;
    } catch (e) {
      return false;
    }
  }

  /// Unsuspend user account
  Future<bool> unsuspendUser(String userId) async {
    try {
      await _firestore.collection('users').doc(userId).update({
        'accountStatus': 'active',
        'suspensionReason': null,
        'suspendedAt': null,
      });
      return true;
    } catch (e) {
      return false;
    }
  }

  /// Delete user (soft delete)
  Future<bool> deleteUser(String userId, {String reason = ''}) async {
    try {
      await _firestore.collection('users').doc(userId).update({
        'accountStatus': 'deleted',
        'deletionReason': reason,
        'deletedAt': FieldValue.serverTimestamp(),
        'displayName': 'Deleted User',
        'bio': 'Account deleted',
        'photos': [],
      });
      return true;
    } catch (e) {
      return false;
    }
  }

  /// Resolve report
  Future<bool> resolveReport(String reportId, String status, {String? adminNotes}) async {
    try {
      await _firestore.collection('reports').doc(reportId).update({
        'status': status,
        'resolvedAt': FieldValue.serverTimestamp(),
        'adminNotes': adminNotes ?? '',
      });
      return true;
    } catch (e) {
      return false;
    }
  }

  /// Get dashboard stats
  Future<Map<String, dynamic>> getDashboardStats() async {
    try {
      final usersSnapshot = await _firestore.collection('users').get();
      final matchesSnapshot = await _firestore.collection('matches').get();
      final reportsSnapshot = await _firestore
          .collection('reports')
          .where('status', isEqualTo: 'pending')
          .get();
      final verificationsSnapshot = await _firestore
          .collection('verifications')
          .where('status', isEqualTo: 'pending')
          .get();

      // Count active users
      final activeUsers = usersSnapshot.docs.where((doc) {
        final status = doc.data()['accountStatus'];
        return status == null || status == 'active';
      }).length;

      // Count verified users
      final verifiedUsers = usersSnapshot.docs.where((doc) {
        final data = doc.data();
        return data['isPhotoVerified'] == true || 
               data['isIDVerified'] == true || 
               data['isPhoneVerified'] == true;
      }).length;

      return {
        'totalUsers': usersSnapshot.size,
        'activeUsers': activeUsers,
        'verifiedUsers': verifiedUsers,
        'totalMatches': matchesSnapshot.size,
        'pendingReports': reportsSnapshot.size,
        'pendingVerifications': verificationsSnapshot.size,
      };
    } catch (e) {
      return {
        'totalUsers': 0,
        'activeUsers': 0,
        'verifiedUsers': 0,
        'totalMatches': 0,
        'pendingReports': 0,
        'pendingVerifications': 0,
      };
    }
  }
}
